src/catalog/src/objects.rs

```diff
@@ -28,6 +28,20 @@ use std::time::Duration;
 
 // Structs used to pass information to outside modules.
 
+/// A trait for representing `Self` as a key (`K`) value (`V`) pair for the purpose of storing this
+/// value durably.
+///
+/// To encode a key-value pair, use [`DurableType::into_key_value`].
+///
+/// To decode a key-value pair, use [`DurableType::from_key_value`].
+pub(crate) trait DurableType<K, V>: Sized {
+    /// Consume and convert `Self` into a key (`K`) value (`V`) pair.
+    fn into_key_value(self) -> (K, V);
+
+    /// Consume and convert a key (`K`) value (`V`) pair back into a `Self` value.
+    fn from_key_value(key: K, value: V) -> Self;
+}
```

Comment:
I like this! What do you think about taking it one step further and making `K` and `V` the protobuf types? For example, right now we go from `Database` -> `(DatabaseKey, DatabaseValue)` -> `(proto::DatabaseKey, proto::DatabaseValue)`. We could eliminate that middle step and go right from `Database` -> `(proto::DatabaseKey, proto::DatabaseValue)`.

This would make `from_key_value(...)` fallible though, so it changes the API a bit, but this be okay? Presumably we're already calling `RustType::from_proto` right before we'd call `from_key_value(...)`, so we should already be in a function that can return an error.

Reply:
If we don't use the protobuf types directly in this trait, can you leave a comment here explaining how this trait is different than the protobuf types or why it doesn't take them directly?

Reply:
Does it make sense to just remove the non proto `XKey` and `XValue` types completely? I think they are only used as an intermediate representation for serialization/deserialization and as fields in the `Transaction` type. It seems like we can remove them and replace all usages with the proto objects directly.

Reply:
> Does it make sense to just remove the non proto XKey and XValue types completely?

100%! Those types only exist to make working with the protos easier, e.g. fewer `Option` fields. I think with this new trait we can remove the `XKey` and `XValue` types entirely and use the proto types directly.

Which btw I would then change the trait so `K` and `V` implement `Data`, e.g.
```
pub(crate) trait DurableType<K, V>: Sized
where
    K: Data,
    V: Data,
{
    /// Consume and convert `Self` into a key (`K`) value (`V`) pair.
    fn into_key_value(self) -> (K, V);

    /// Consume and convert a key (`K`) value (`V`) pair back into a `Self` value.
    fn from_key_value(key: K, value: V) -> Self;
}
```

Reply:
What about getting rid of the trait completely and doing something like 
```Rust
impl RustType<(proto::DatabaseKey, proto::DatabaseValue)> for Database {
    ...
}
```

> 100%! Those types only exist to make working with the protos easier, e.g. fewer `Option` fields. I think with this new trait we can remove the `XKey` and `XValue` types entirely and use the proto types directly.

Oh right, I forgot about the `Option`s. That will probably make be very annoying to deal with in `Transaction`s, so we probably need to keep around the `XKey` and `XValue` types for that.

Reply:
Oy, then there's going to be a ton of duplicate code between methods like

```Rust
impl RustType<(proto::DatabaseKey, proto::DatabaseValue)> for Database {
    ...
}
```
and
```Rust
impl RustType<proto::DatabaseKey> for DatabaseKey {
    ...
}
```

I think I'm just going to leave it as is and add a comment and issue like you described.